sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/complexTypesSuite.scala

```diff
@@ -0,0 +1,499 @@
+          SimplifyBinaryComparison,
+          SimplifyCreateStructOps,
+          SimplifyCreateArrayOps,
+          SimplifyCreateMapOps) :: Nil
+  }
+
+  val idAtt = ('id).long.notNull
+
+  lazy val baseOptimizedPlan = Range(1L, 1000L, 1, Some(2), idAtt :: Nil)
+
+  val idRef = baseOptimizedPlan.output.head
+
+  implicit class ComplexTypeDslSupport(e : Expression) {
+    def getStructField(f : String): GetStructField = {
+      assert(e.resolved)
+      assert(e.dataType.isInstanceOf[StructType])
+      val structType = e.dataType.asInstanceOf[StructType]
+      val ord = structType.fieldNames.indexOf(f)
+      assert(-1 != ord)
+      GetStructField(e, ord, Some(f))
+    }
+
+    def getArrayStructField(f : String) : Expression = {
+      assert(e.resolved)
+      assert(e.dataType.isInstanceOf[ArrayType])
+      val arrType = e.dataType.asInstanceOf[ArrayType]
+      assert(arrType.elementType.isInstanceOf[StructType])
+      val structType = arrType.elementType.asInstanceOf[StructType]
+      val ord = structType.fieldNames.indexOf(f)
+      assert(-1 != ord)
+      GetArrayStructFields(e, structType(ord), ord, 1, arrType.containsNull)
+    }
+
+    def getArrayItem(i : Int) : GetArrayItem = {
+      assert(e.resolved)
+      assert(e.dataType.isInstanceOf[ArrayType])
+      GetArrayItem(e, Literal(i))
+    }
+
+    def getMapValue(k : Expression) : Expression = {
+      assert(e.resolved)
+      assert(e.dataType.isInstanceOf[MapType])
+      val mapType = e.dataType.asInstanceOf[MapType]
+      assert(k.dataType == mapType.keyType)
+      GetMapValue(e, k)
+    }
+
+    def addCaseWhen( cond : Expression, v : Expression) : Expression = {
+      assert(e.resolved)
+      assert(e.isInstanceOf[CaseWhen])
+      assert(cond.dataType == BooleanType)
+      assert(v.dataType == e.dataType)
+      val CaseWhen(branches, default) = e
+      CaseWhen( branches :+ (cond, v), default)
+    }
+  }
+
+  test("explicit") {
+    val rel = baseOptimizedPlan.select(
+      CreateNamedStruct("att" :: idRef :: Nil).getStructField("att") as "outerAtt"
+   )
+
+    assertResult(StructType(StructField("outerAtt", LongType, nullable = false) :: Nil))(rel.schema)
+
+    val optimized = Optimize execute rel
+
+    val expected = baseOptimizedPlan.select(idRef as "outerAtt")
+
+    comparePlans(optimized, expected)
+  }
+
+  ignore("explicit - deduced att name") {
+    val rel = baseOptimizedPlan.select(
+      CreateNamedStruct("att" :: idRef :: Nil).getStructField("att")
+   )
+    assertResult(StructType(
+      StructField("named_struct(att, id AS `att`).att", LongType, nullable = false) :: Nil
+    ))(rel.schema)
+
+    val optimized = Optimize execute rel
+
+    val expected = baseOptimizedPlan.select(idRef as "named_struct(att, id AS `att`).att")
+
+    comparePlans(optimized, expected)
+  }
+
+  test("collapsed") {
+    val rel = baseOptimizedPlan.select(
+      CreateNamedStruct("att" :: idRef :: Nil) as "struct1"
+    )
+    assertResult(

```

Comment:
basic idea was to test that this piece of expression maintains the attribute name through optimization phases. can be removed if you think it's unnecessary here.